src/hypofuzz/dashboard.py

```diff
@@ -149,10 +216,65 @@ async def serve_app(app: Any, host: str, port: str) -> None:
     await serve(app, config)
 
 
+async def handle_event(receive_channel: MemoryReceiveChannel) -> None:
+    async for event_type, args in receive_channel:
+        if event_type == "save":
+            key, value = args
+            assert value is not None
+            if key.endswith(reports_key):
+                report = json.loads(value)
+                REPORTS[report["nodeid"]].add(report)
+                await broadcast_event("save", report)
+            if key.endswith(metadata_key):
+                metadata = json.loads(value)
+                METADATA[metadata["nodeid"]] = metadata
+                await broadcast_event("metadata", metadata)
+
+        if event_type == "delete":
+            key, value = args
+
+            if key.endswith(reports_key):
+                key = key.rstrip(reports_key)
+                if value is not None:
+                    report = json.loads(value)
+                    REPORTS[report["nodeid"]].remove(report)
+                    await broadcast_event("delete", report)
+                else:
+                    # debounce each key, to relieve db read pressure for this expensive
+                    # re-scan. Deletions aren't important for correctness, just
+                    # performance / memory.
+                    if time.time() - refreshed_at[key] > delete_debounce:
+                        reports = SortedList(
+                            list(get_db().fetch_reports(key)),
+                            key=lambda r: r["elapsed_time"],
+                        )
+                        if reports:
+                            node_id = reports[0]["nodeid"]
+                            REPORTS[node_id] = reports
+                            await broadcast_event("refresh", reports)
+                        refreshed_at[key] = time.time()
+
+            # We're only keeping a reference to the latest metadata, so we don't need
+            # to bother handling deletion events for it, just save.
+
+
 async def run_dashboard(port: int, host: str) -> None:
+    send_channel, receive_channel = trio.open_memory_channel[ListenerEventT](math.inf)
+    db = get_db()
+    db._db.add_listener(
+        # DirectoryBasedExampleDatabase sends events from a background thread (via watchdog).
+        partial(
+            trio.from_thread.run_sync,
+            send_channel.send_nowait,  # type: ignore
+            trio_token=trio.lowlevel.current_trio_token(),
+        )
+        # two layers of unwrapping: db is a HypofuzzDatabase(BackgroundWriteDatabse(<actual database here>))
+        if isinstance(db._db._db, DirectoryBasedExampleDatabase)  # type: ignore
+        else send_channel.send_nowait
+    )
```

Comment:
I'd be sad to break user-defined databases which run the listener in a thread but don't exactly match this pattern; fortunately I think we can handle that with a listener function that checks where it's called:
```suggestion
    token = trio.lowlevel.current_trio_token()

    def send_nowait_from_anywhere(msg: object) -> None:
        # DirectoryBasedExampleDatabase sends events from a background thread (via watchdog),
        # so we need to support sending from anywhere, i.e. whether or not the calling thread
        # has any Trio state.  We can do that with the following branch: 
        try:
            trio.lowlevel.current_task()
        except RuntimeError:
            trio.from_thread.run_sync(send_channel.send_nowait, msg, trio_token=token)
        else:
            send_channel.send_nowait(msg)

    db = get_db()
    db._db.add_listener(send_nowait_from_anywhere)
```

Reply:
oops, apparently I should have typed `msg: ListenerEventT` 😅

Reply:
sweet ty